T 20 loading admin data

Ali-Alkhars opened the pull request:
Hello everyone! T-20 is done, which was about **loading choice boards data from the database** instead of the hardcoded data.

### Here are the main things to look at:
- How `admin_choice_boards.dart` changed
- Draggable list has been refactored into three files: `categories.dart`, `category.dart` and `category_item.dart`
- The _Retrieving data functions_ methods I added to `firebase_functions.dart`. You'll find them at the very end
- Added unit tests only for one of the three methods I wrote in `firebase_functions.dart` because the others use the cache and for some reason, flutter testing just ignores anything that uses the cache!
- Try to create data and do whatever you wish with it, and then check if it's consistent with the database

Ali-Alkhars committed:
Allow dynamic loading of choice boards data

- Add a firebase function to retrieve the user's data and convert it into DraggableList
- Edit the hardcoded data and test data to have all the fields as in the database
- Adjust Wrapper and the side menu to accommodate the changes

Ali-Alkhars committed:
Fix loading choice boards data

- Used a LoadingMixin to load async data
- Adjusted some files to accommodate recent changes
- Removed debug prints

Ali-Alkhars committed:
Make data get loaded in the correct rank

- Create a widget for a loading screen
- Make editing or deleting an item refresh the page to see the changes
- Add comments

Ali-Alkhars committed:
Do appropriate error handling

- Change updateCategoryRank name into updateCategoryItemsRank to avoid confusion
- Remove a misleading comment in delete item button

Ali-Alkhars committed:
Make choice boards page take data from the cache if not connected to the internet

- Added functions in firebase_functions to facilitate storing data in the cache
- Added a class "Categories" in draggable_list.dart which is a list of DraggableList
- Added methods in draggable_list.dart to convert data to/from json to be able to cache the data

Ali-Alkhars committed:
Refactor draggable_list.dart

- Refactor DraggableList to Category in the same file
- Refactor DraggableListItem to CategoryItem in a new file
- Refactor Categories to a new file
- Add comments to all the files
- Edit the files which use those datatypes to use the new names

Ali-Alkhars committed:
Fix test errors occuring due to recent changes

- Make sure that newly made Categories model is used where necessary

Ali-Alkhars committed:
Add instructions texts

- Change error messages to ask users to check their internet connection
- Add messages for when choice boards is empty or a category is empty

Ali-Alkhars committed:
Merge branch 'master' into T-20-Loading-Admin-Data

Ali-Alkhars committed:
Merge branch 'master' into T-20-Loading-Admin-Data

Ali-Alkhars committed:
Add unit tests for downloadUserCategories

Ali-Alkhars committed:
delete duplicated test file

absirgue reviewed (changes requested):
Hey Monsieur,

Thanks for the work, love the refactoring and amazing work loading the categories. It makes everything finally work together on the admin side and it feels awesome!

I just have 2 comments from the manual testing I did:
1. It takes a bit of time for the app to realize I'm not connected to the internet and load from cache. Maybe instead of waiting for the database to throw an error you can wait for either the database to be thrown or a certain time span to pass. Like this, we would fully benefit from the caching and also handle situations of simply low wifi.
2. Editing the name of a category isn't blocked but since I'm not connected to the wifi it gives me a very very long loading screen. It seems like the rereading of item is blocked when off wifi, can you block the editing/deleting methods? 
=> you can maybe find how to build a method that check whether or not the app is connected to wifi here: https://stackoverflow.com/questions/49648022/check-whether-there-is-an-internet-connection-available-on-flutter-app

Finally, I was checking test coverage and the adding of the toJSON() and fromJSON() methods in most models lowers their test coverage by a lot since these two methods combined form like 70% of the nb of lines of each of these files. Could you write unit tests for them that check if a category/item object is nicely translated to a JSON object and if a JSON object is correctly translated to a category/item object?


Thanks!

AymenBerb reviewed (approved):
pull main before mege

Ali-Alkhars committed:
Write tests for Categories, Category and CategoryItem models

- Add equality operator overrides for those models

Ali-Alkhars committed:
Add internet connectivity listener

- Make admin choice boards buttons unaccessible if the user is not connected to the internet
- Fix all tests that fail because of the changes

Ali-Alkhars committed:
Merge branch 'master' into T-20-Loading-Admin-Data

absirgue reviewed (approved):
Fantastic work!
